contracts/RemovalQueue.sol

```diff
@@ -0,0 +1,177 @@
+   * @dev The removal is added to the Enumberable Set that maps to the year of its vintage.
+   * @param removalToInsert a new removal to insert.
+   * @return bool true if success, false otherwise.
+   */
+  function insertRemovalByVintage(
+    RemovalQueueByVintage storage removalQueue,
+    uint256 removalToInsert
+  ) internal returns (bool) {
+    uint16 vintageOfRemoval = removalToInsert.vintage();
+    if (isRemovalQueueEmpty(removalQueue)) {
+      removalQueue.earliestYear = vintageOfRemoval;
+      removalQueue.latestYear = vintageOfRemoval;
+    } else if (vintageOfRemoval < removalQueue.earliestYear) {
+      removalQueue.earliestYear = vintageOfRemoval;
+    } else if (vintageOfRemoval > removalQueue.latestYear) {
+      removalQueue.latestYear = vintageOfRemoval;
+    }
+    return removalQueue.queueByVintage[vintageOfRemoval].add(removalToInsert);
+  }
+
+  /**
+   * @notice Removes a removal from the queue.
+   * @dev Removes the removal from the Enumerable Set that corresponds to its vintage.
+   * @param removalQueue the queue to search through.
+   * @param removalToRemove the removal to remove.
+   * @return bool true if success, false otherwise.
+   */
+  function removeRemoval(
+    RemovalQueueByVintage storage removalQueue,
+    uint256 removalToRemove
+  ) internal returns (bool) {
+    uint16 vintageOfRemoval = removalToRemove.vintage();
+    require(
+      removalQueue.queueByVintage[vintageOfRemoval].remove(removalToRemove),
+      "Market: failed to remove correct removal"
+    );
+    // If all removals were removed, check to see if there are any updates to the struct we need to make.
+    if (isRemovalQueueEmptyForVintage(removalQueue, vintageOfRemoval)) {
+      if (removalQueue.earliestYear == removalQueue.latestYear) {
+        // If there was only one year remaining, clear the values for latest and earliest years.
+        removalQueue.earliestYear = _DEFAULT_EARLIEST_YEAR;
+        removalQueue.latestYear = _DEFAULT_LATEST_YEAR;
+      } else if (vintageOfRemoval == removalQueue.earliestYear) {
+        // If this was the earliest year, find the new earliest year and update the struct.
+        for (
+          uint16 currentYear = removalQueue.earliestYear + 1;
+          currentYear <= removalQueue.latestYear;
+          currentYear++
+        ) {
+          if (removalQueue.queueByVintage[currentYear].length() > 0) {
+            removalQueue.earliestYear = currentYear;
+            break;
+          }
+        }
+      } else if (vintageOfRemoval == removalQueue.latestYear) {
+        // If this was the latest year, find the new latest year and update the struct.
+        for (
+          uint16 currentYear = removalQueue.latestYear - 1;
+          currentYear >= removalQueue.earliestYear;
+          currentYear--
+        ) {
+          if (removalQueue.queueByVintage[currentYear].length() > 0) {
+            removalQueue.latestYear = currentYear;
+            break;
+          }
+        }
```

Comment:
is this a good use case for this? https://docs.openzeppelin.com/contracts/4.x/api/utils#Arrays-findUpperBound-uint256---uint256-

Reply:
I don't think so, since we're not iterating through an array, it's a mapping. Unless I'm misunderstanding something.